File: leetcode_solved/leetcode_0303_Range_Sum_Query_-_Immutable.cpp
/**
 * AC:
 * Runtime: 28 ms, faster than 93.42% of C++ online submissions for Range Sum Query - Immutable.
 * Memory Usage: 17.1 MB, less than 96.55% of C++ online submissions for Range Sum Query - Immutable.
 */
class NumArray {
public:
    NumArray(vector<int>& nums):record(nums.size() + 1, 0) {
        partial_sum(nums.begin(), nums.end(), record.begin() + 1);
    }
    
    int sumRange(int i, int j) {
  		return record[j + 1] - record[i];      
    }

private:
	vector<int> record;    
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(i,j);
 */

File: leetcode_solved/leetcode_0303_Range_Sum_Query_-_Immutable.java
// AC: Runtime: 6 ms, faster than 100.00% of Java online submissions for Range Sum Query - Immutable.
// Memory Usage: 44.1 MB, less than 13.72% of Java online submissions for Range Sum Query - Immutable.
// prefix sum
// T:O(1), S:O(n)
//
class NumArray {
    private int[] prefixSum;

    public NumArray(int[] nums) {
        prefixSum = new int[nums.length + 1];
        prefixSum[0] = 0;
        int tempSum = 0, pos = 1;
        for (int i: nums) {
            tempSum += i;
            prefixSum[pos++] = tempSum;
        }
    }

    public int sumRange(int left, int right) {
        return prefixSum[right + 1] - prefixSum[left];
    }
}